### PR TITLE
Qualify pegged/peg.d as @safe and when possible pure nothrow @nogc

### PR DESCRIPTION
**Pull request by nordlow**

### CODE REVIEW

**Comment by nordlow** [comment]
Goes green locally now, @John-Colvin.

File: pegged/parser.d [hunk 1]
```diff
@@ -137,18 +137,20 @@ public import pegged.peg;
...
     enum name = "Pegged";
-    static ParseTree delegate(ParseTree)[string] before;
-    static ParseTree delegate(ParseTree)[string] after;
-    static ParseTree delegate(ParseTree)[string] rules;
-    static this()
+    static ParseTree delegate(ParseTree) @safe [string] before;
+    static ParseTree delegate(ParseTree) @safe [string] after;
+    static ParseTree delegate(ParseTree) @safe [string] rules;
+    static this() @trusted
     {
```
Comment (veelo) [inline pegged/parser.d:154]: Interesting. Any idea why this can't be `@safe`?

**Review by veelo — COMMENTED** [review]
Thanks for doing this!

**Comment by veelo** [comment]
It seems to me that the CI failures are due to deficiencies in moderately recent (2 years old) druntime.

I am thinking of determining the release at which this starts working a bit more precisely, putting that in a [`toolchainRequirements` section](https://dub.pm/package-format-json.html#toolchain-requirements), and releasing this with a mayor version bump. Would that work for Symmetry @John-Colvin? What is the oldest D frontend release that you depend on?

File: pegged/parser.d (hunk 1 above)
Comment (nordlow) [reply pegged/parser.d:154]: toDelegate is @system.

File: pegged/peg.d [hunk 2]
```diff
@@ -1741,7 +1746,7 @@ template or(rules...) if (rules.length > 0)
                 start += len + names[i].length + 4;
             }
         }
-        orErrorString = cast(string)(errString[0..$-4]);
+		() @trusted { orErrorString = cast(string)(errString[0..$-4]); } ();
```
[the remark this answers is not in the document]
Comment (nordlow) [reply pegged/peg.d:1750]: Idem? Do you mean it should be replaced with `string.reserve`?

**Comment by nordlow** [comment]
Hitting 2.091-version compiler bugs aswell at https://ci.appveyor.com/project/PhilippeSigaud/pegged/builds/43483820/job/3503ke4jawrotjej#L83.

**Comment by John-Colvin** [comment]
We only care about the compilers we actively support (@nordlow the ones our setup scripts install)

File: pegged/peg.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (veelo) [reply pegged/peg.d:1750]: Yeah, see https://github.com/PhilippeSigaud/Pegged/pull/311#discussion_r867999676.

**Comment by nordlow** [comment]
In appveyor.yml, I suggest we remove support for dmd 2.087.1 and bump 2.090.1 to 2.091.0 having @safe `AA.keys`. That way we can remove the versioning added.

**Comment by nordlow** [comment]
> 
We only care about the compilers we actively support (@nordlow the ones our setup scripts install)

The oldest compiler version _we_ need to support is 2.099.1. So to which compiler version less or equal to that do you want to bump Pegged CI to?

**Review by veelo — CHANGES_REQUESTED** [review]
Can you please react to https://github.com/PhilippeSigaud/Pegged/pull/311#discussion_r867999676? I think it fits into this PR. If you don't want to do it then that's fine, and I'll see if I can do a follow-up PR.

**Comment by nordlow** [comment]
> Can you please react to [#311 (comment)](https://github.com/PhilippeSigaud/Pegged/pull/311#discussion_r867999676)? I think it fits into this PR. If you don't want to do it then that's fine, and I'll see if I can do a follow-up PR.

Sorry, missed that. I'll have look.

**Comment by veelo** [comment]
> So to which compiler version less or equal to <= 2.099.1 do you want to bump `matrix`-entries in `appveyor.yml` to, @veelo?

I think we should try to support as far back as we reasonably can, so like you suggest:

> removing support for dmd 2.087.1 and bump 2.090.1 to 2.091.0

but then also make that explicit with a [`toolchainRequirements` section](https://dub.pm/package-format-json.html#toolchain-requirements) section in the Dub configuration. 

I'll have to think about whether that demands a major version bump, since technically that would brake the build for people that don't update their compiler regularly. I'll ask the forum.

**Comment by nordlow** [comment]
Correction .keys became @safe version in 2.098.0 as per https://dlang.org/changelog/2.098.0.html#bugfix-list. Gonna update.

**Comment by veelo** [comment]
> I'll ask the forum.

https://forum.dlang.org/post/pxyqgiyosabquyjzwxex@forum.dlang.org

File: pegged/dynamic/peg.d [hunk 3]
```diff
@@ -309,7 +311,7 @@ Dynamic or(T...)(T rules)
                 start += len + names[i].length + 4;
             }
         }
-        orErrorString = cast(string)(errString[0..$-4]);
+		() @trusted { orErrorString = cast(string)(errString[0..$-4]); } ();
```
[the remark this answers is not in the document]
Comment (nordlow) [reply pegged/dynamic/peg.d:315]: I'd be grateful if you can deal with this issue. Ok?

**Comment by nordlow** [comment]
> I cannot release this yet, since parser.d is a generated parser (Pegged generates its own parser). Whenever parser.d is regenerated, the @safe annotations will be reverted from it. Pegged will need to be modified to generate @safe parsers.

Can you give me a sample command line call that does this? Then I can dig into having the generated parser be `@safe` as well.

**Comment by nordlow** [comment]
Are you saying the parser generated from

```d
cd pegged/dev
rdmd -I../.. -I../../examples/peggedgrammar/src -I../../examples/misc/src regenerate.d
```

lacks @safe-qualifiers?

**Comment by nordlow** [comment]
Can you please bump https://code.dlang.org/packages/pegged tag aswell to 0.4.6? It's still at 0.4.5.

**Comment by veelo** [comment]
> Is there anything more you need help with?

Seriously? There is a lot to do actually, but most of it requires a deeper dive. We can talk about those if you are interested. 

Otherwise, I'd really like an automated check on PRs to make sure the committed `parser.d` is the same as one that is generated.

File: pegged/dynamic/peg.d (hunk 3 above)
[the remark this answers is not in the document]
Comment (veelo) [reply pegged/dynamic/peg.d:315]: #312

**Comment by veelo** [comment]
> Can you please bump https://code.dlang.org/packages/pegged tag aswell to 0.4.6? It's still at 0.4.5.

To my knowledge, it should update automatically. It probably scans for new tags every so many hours.

**Comment by nordlow** [comment]
> To my knowledge, it should update automatically. It probably scans for new tags every so many hours.

Still not bump to 0.4.6 at https://code.dlang.org/packages/pegged. Can you bump it yourself?

**Comment by veelo** [comment]
No, I don't even have an account. I think that one is still with @PhilippeSigaud. I am wondering if it is because the other [releases](https://github.com/PhilippeSigaud/Pegged/releases) have a commit signature icon, but I don't know why the latest doesn't.

**Comment by nordlow** [comment]
> No, I don't even have an account. I think that one is still with @PhilippeSigaud. I am wondering if it is because the other [releases](https://github.com/PhilippeSigaud/Pegged/releases) have a commit signature icon, but I don't know why the latest doesn't.

Do you have any clues as to why, @John-Colvin @skoppe?

**Comment by veelo** [comment]
I don't think the signature has anything to do with it, other packages that *did* update in the Dub registry within the last 16 hours don't have that icon either. I've checked the syntax of the tag multiple times, should be no problems there. I made sure that the pre-release checkbox is unchecked. The Dub docs say that tags are scanned [about twice an hour](https://dub.pm/publish.html). This is not my first release, and I don't think I did anything different from last time.

**Comment by nordlow** [comment]
> I don't think the signature has anything to do with it, other packages that _did_ update in the Dub registry within the last 16 hours don't have that icon either. I've checked the syntax of the tag multiple times, should be no problems there. I made sure that the pre-release checkbox is unchecked. The Dub docs say that tags are scanned [about twice an hour](https://dub.pm/publish.html). This is not my first release, and I don't think I did anything different from last time.

FYI, @CyberShadow.

**Comment by CyberShadow** [comment]
I remember that the updater used to sometimes get stuck. Sonke would come in and restart it. However, I just tried to update my own package and it succeeded, so I don't think it's that in this case. It seems to have trouble with this specific package.

A package admin could go to https://code.dlang.org/my_packages/pegged and perhaps that page would show an error message.

@s-ludwig Seems that we need your help again; also, if there's anyone else who could help here, perhaps they should be listed on https://wiki.dlang.org/People#Points_of_contact .

**Comment by nordlow** [comment]
Further note that https://github.com/PhilippeSigaud has been inactive for the last 1 year.

**Comment by veelo** [comment]
Maybe it's time to transfer Pegged to https://github.com/dlang-community? I'm fine with continuing as a maintainer of Pegged. At least then there is a group of people that are able to pick up the ball it case it is dropped. Is there a shared dub account for those projects? What is the procedure for such a transfer?

**Comment by nordlow** [comment]
> Maybe it's time to transfer Pegged to https://github.com/dlang-community? I'm fine with continuing as a maintainer of Pegged. At least then there is a group of people that are able to pick up the ball it case it is dropped. Is there a shared dub account for those projects? What is the procedure for such a transfer?

I vote for that. Ping, @John-Colvin.

**Comment by skoppe** [comment]
> I remember that the updater used to sometimes get stuck

Yes, I have had this happen repeatably the last year. Sometimes it works, sometimes it doesn't. The page says its processing, but nothing happens (for days) unless I manually hit the button, all the while other packages are just updating.

**Comment by veelo** [comment]
It finally got through now.

**Comment by nordlow** [comment]
> It finally got through now.

Thank you. Was it thanks to somebody manually synching code.dlang.org?

**Comment by veelo** [comment]
> Was it thanks to somebody manually refreshing the tag on code.dlang.org?

Don't know. There was an issue with an [expired certificate](https://forum.dlang.org/post/t5ag7e$2n95$1@digitalmars.com), but I doubt it is related because other packages had no issue.

**Pull request merged**